lib/apiservers/service/restapi/handlers/vch_list_get.go

```diff
@@ -0,0 +1,209 @@
+		d.ComputeResourcePath = *computeResource
+	}
+
+	return &d
+}
+
+func validateTarget(ctx context.Context, d *data.Data) (*validate.Validator, error) {
+	if err := d.HasCredentials(); err != nil {
+		return nil, fmt.Errorf("Invalid Credentials: %s", err)
+	}
+
+	validator, err := validate.NewValidator(ctx, d)
+	if err != nil {
+		return nil, fmt.Errorf("Validation Error: %s", err)
+	}
+	// If dc is not set, and multiple datacenter is available, vic-machine ls will list VCHs under all datacenters.
+	validator.AllowEmptyDC()
+
+	_, err = validator.ValidateTarget(ctx, d)
+	if err != nil {
+		return nil, fmt.Errorf("Target validation failed: %s", err)
+	}
+	_, err = validator.ValidateCompute(ctx, d, false)
+	if err != nil {
+		return nil, fmt.Errorf("Compute resource validation failed: %s", err)
+	}
+
+	return validator, nil
+}
+
+// Copied from list.go, and appears to be present other places. TODO: deduplicate
+func upgradeStatusMessage(ctx context.Context, vch *vm.VirtualMachine, installerVer *version.Build, vchVer *version.Build) string {
+	if sameVer := installerVer.Equal(vchVer); sameVer {
+		return "Up to date"
+	}
+
+	upgrading, err := vch.VCHUpdateStatus(ctx)
+	if err != nil {
+		return fmt.Sprintf("Unknown: %s", err)
+	}
+	if upgrading {
+		return "Upgrade in progress"
+	}
+
+	canUpgrade, err := installerVer.IsNewer(vchVer)
+	if err != nil {
+		return fmt.Sprintf("Unknown: %s", err)
+	}
+	if canUpgrade {
+		return fmt.Sprintf("Upgradeable to %s", installerVer.ShortVersion())
+	}
+
+	oldInstaller, err := installerVer.IsOlder(vchVer)
+	if err != nil {
+		return fmt.Sprintf("Unknown: %s", err)
+	}
+	if oldInstaller {
+		return fmt.Sprintf("VCH has newer version")
+	}
+
+	// can't get here
+	return "Invalid upgrade status"
+}
+
+func vchsToModels(ctx context.Context, vchs []*vm.VirtualMachine, executor *management.Dispatcher) []*models.VCHListItem {
+	installerVer := version.GetBuild()
+	payload := make([]*models.VCHListItem, 0)
+	for _, vch := range vchs {
+		var version *version.Build
+		var dockerHost string
+		var adminPortal string
+		if vchConfig, err := executor.GetNoSecretVCHConfig(vch); err == nil {
+			version = vchConfig.Version
+
+			if public := vchConfig.ExecutorConfig.Networks["public"]; public != nil {
+				if public_ip := public.Assigned.IP; public_ip != nil {
+					var docker_port string
+					if !vchConfig.HostCertificate.IsNil() {
+						docker_port = fmt.Sprintf("%d", opts.DefaultTLSHTTPPort)
+					} else {
+						docker_port = fmt.Sprintf("%d", opts.DefaultHTTPPort)
+					}
+
+					dockerHost = fmt.Sprintf("%s:%s", public_ip, docker_port)
+					adminPortal = fmt.Sprintf("https://%s:2378", public_ip)
+				}
+			}
+		}
+
+		name := path.Base(vch.InventoryPath)
+
+		model := &models.VCHListItem{ID: vch.Reference().Value, Name: name}
+
+		if version != nil {
+			model.Version = version.ShortVersion()
+			model.UpgradeStatus = upgradeStatusMessage(ctx, vch, installerVer, version)
+		}
+
+		if adminPortal != "" {
```

Comment:
the default values of AdminPortal and DockerHost is "" anyway so I think you could just assign them without checking